src/uct/sm/mm/base/mm_iface.c

```diff
@@ -222,81 +227,128 @@ static inline ucs_status_t uct_mm_iface_process_recv(uct_mm_iface_t *iface,
+}
 
-    /* check the memory pool to make sure that there is a new descriptor available */
-    if (ucs_unlikely(iface->last_recv_desc == NULL)) {
-        UCT_TL_IFACE_GET_RX_DESC(&iface->super.super, &iface->recv_desc_mp,
-                                 iface->last_recv_desc, return 0);
+static inline unsigned
+uct_mm_iface_poll_fifo(uct_mm_iface_t *iface)
+{
+    if (!uct_mm_iface_fifo_has_new_data(iface)) {
+        return 0;
     }
 
-    read_index = iface->read_index;
-    read_index_loc = (read_index & iface->fifo_mask);
-    /* the fifo_element which the read_index points to */
-    read_index_elem = UCT_MM_IFACE_GET_FIFO_ELEM(iface, iface->recv_fifo_elems,
-                                                 read_index_loc);
+    /* read from read_index_elem */
+    ucs_memory_cpu_load_fence();
+    ucs_assert(iface->read_index <= iface->recv_fifo_ctl->head);
 
-    /* check the read_index to see if there is a new item to read (checking the owner bit) */
-    if (((read_index >> iface->fifo_shift) & 1) == ((read_index_elem->flags) & 1)) {
+    uct_mm_iface_process_recv(iface, iface->read_index_elem);
 
-        /* read from read_index_elem */
-        ucs_memory_cpu_load_fence();
-        ucs_assert(iface->read_index <= iface->recv_fifo_ctl->head);
+    /* raise the read_index */
+    iface->read_index++;
 
-        status = uct_mm_iface_process_recv(iface, read_index_elem);
-        if (status != UCS_OK) {
-            /* the last_recv_desc is in use. get a new descriptor for it */
-            UCT_TL_IFACE_GET_RX_DESC(&iface->super.super, &iface->recv_desc_mp,
-                                     iface->last_recv_desc, ucs_debug("recv mpool is empty"));
-        }
+    /* the next fifo_element which the read_index points to */
+    iface->read_index_elem =
+        UCT_MM_IFACE_GET_FIFO_ELEM(iface, iface->recv_fifo_elems,
+                                   (iface->read_index & iface->fifo_mask));
+
+    uct_mm_progress_fifo_tail(iface);
 
-        /* raise the read_index. */
-        iface->read_index++;
+    return 1;
+}
 
-        uct_mm_progress_fifo_tail(iface);
+static inline void
+uct_mm_iface_fifo_window_adjust(uct_mm_iface_t *iface,
+                                unsigned fifo_poll_count)
+{
+    /* Adjust window if was completed at least one RX operation
+     * during polling */
+    if (fifo_poll_count == 0) {
+        return;
+    }
 
-        return 1;
+    if (fifo_poll_count == iface->fifo_wnd) {
+        if (iface->fifo_prev_wnd_cons) {
+            /* Increase FIFO window size if it was fully consumed
+             * during the previous iface progress call */
+            iface->fifo_wnd = ucs_min(iface->fifo_wnd +
+                                      UCT_MM_IFACE_FIFO_AI_VALUE,
+                                      iface->config.fifo_max_wnd);
+        } else {
+            iface->fifo_prev_wnd_cons = 1;
+        }
     } else {
-        return 0;
+        iface->fifo_wnd = ucs_max(iface->fifo_wnd /
```

Comment:
mayb better check this at first and return?
`if (poll_count < fifo_wnd) { update wnd; return}`

Reply:
yes, it is better
done